Mathlib/Geometry/Manifold/ContMDiffMFDeriv.lean

```diff
@@ -0,0 +1,625 @@
+  [NormedSpace 𝕜 F₂]
+  -- declare functions, sets, points and smoothness indices
+  {f f₁ : M → M'}
+  {s s₁ t : Set M} {x : M} {m n : ℕ∞}
+
+-- Porting note: section about deducing differentiability from smoothness moved to
+-- `Geometry.Manifold.`MFDeriv`
+
+/-! ### The derivative of a smooth function is smooth -/
+
+section mfderiv
+
+/-- The function that sends `x` to the `y`-derivative of `f(x,y)` at `g(x)` is `C^m` at `x₀`,
+where the derivative is taken as a continuous linear map.
+We have to assume that `f` is `C^n` at `(x₀, g(x₀))` for `n ≥ m + 1` and `g` is `C^m` at `x₀`.
+We have to insert a coordinate change from `x₀` to `x` to make the derivative sensible.
+This result is used to show that maps into the 1-jet bundle and cotangent bundle are smooth.
+`cont_mdiff_at.mfderiv_id` and `ContMDiffAt.mfderiv_const` are special cases of this.
+
+This result should be generalized to a `ContMDiffWithinAt` for `mfderivWithin`.
+If we do that, we can deduce `ContMDiffOn.contMDiffOn_tangentMapWithin` from this.
+-/
+protected theorem ContMDiffAt.mfderiv {x₀ : N} (f : N → M → M') (g : N → M)
+    (hf : ContMDiffAt (J.prod I) I' n (Function.uncurry f) (x₀, g x₀)) (hg : ContMDiffAt J I m g x₀)
+    (hmn : m + 1 ≤ n) :
+    ContMDiffAt J 𝓘(𝕜, E →L[𝕜] E') m
+      (inTangentCoordinates I I' g (fun x => f x (g x)) (fun x => mfderiv I I' (f x) (g x)) x₀)
+      x₀ := by
+  have h4f : ContinuousAt (fun x => f x (g x)) x₀ :=
+    ContinuousAt.comp_of_eq hf.continuousAt (continuousAt_id.prod hg.continuousAt) rfl
+  have h4f := h4f.preimage_mem_nhds (extChartAt_source_mem_nhds I' (f x₀ (g x₀)))
+  have h3f := contMDiffAt_iff_contMDiffAt_nhds.mp (hf.of_le <| (self_le_add_left 1 m).trans hmn)
+  have h2f : ∀ᶠ x₂ in 𝓝 x₀, ContMDiffAt I I' 1 (f x₂) (g x₂) := by
+    refine' ((continuousAt_id.prod hg.continuousAt).tendsto.eventually h3f).mono fun x hx => _
+    exact hx.comp (g x) (contMDiffAt_const.prod_mk contMDiffAt_id)
+  have h2g := hg.continuousAt.preimage_mem_nhds (extChartAt_source_mem_nhds I (g x₀))
+  have :
+    ContDiffWithinAt 𝕜 m
+      (fun x =>
+        fderivWithin 𝕜
+          (extChartAt I' (f x₀ (g x₀)) ∘ f ((extChartAt J x₀).symm x) ∘ (extChartAt I (g x₀)).symm)
+          (range I) (extChartAt I (g x₀) (g ((extChartAt J x₀).symm x))))
+      (range J) (extChartAt J x₀ x₀) := by
+    rw [contMDiffAt_iff] at hf hg
+    simp_rw [Function.comp, uncurry, extChartAt_prod, LocalEquiv.prod_coe_symm,
+      ModelWithCorners.range_prod] at hf ⊢
+    refine' ContDiffWithinAt.fderivWithin _ hg.2 I.unique_diff hmn (mem_range_self _) _
+    · simp_rw [extChartAt_to_inv]; exact hf.2
+    · rw [← image_subset_iff]
+      rintro _ ⟨x, -, rfl⟩
+      exact mem_range_self _
+  have :
+    ContMDiffAt J 𝓘(𝕜, E →L[𝕜] E') m
+      (fun x =>
+        fderivWithin 𝕜 (extChartAt I' (f x₀ (g x₀)) ∘ f x ∘ (extChartAt I (g x₀)).symm) (range I)
+          (extChartAt I (g x₀) (g x)))
+      x₀ := by
```

Comment:
Very minor, but I was confused by the syntax here. Perhaps it would be more readable with `x₀` at the end of the previous line?